WMF Framework/Remote Notifications/RemoteNotificationsAPIController.swift

```diff
@@ -0,0 +1,203 @@
+        let query: Query?
+        let error: ResultError?
+
+        var succeeded: Bool {
+            return query?.markAsRead?.result == .success
+        }
+
+        struct Query: Decodable {
+            let markAsRead: MarkedAsRead?
+
+            enum CodingKeys: String, CodingKey {
+                case markAsRead = "echomarkread"
+            }
+        }
+        struct MarkedAsRead: Decodable {
+            let result: Result?
+        }
+        enum Result: String, Decodable {
+            case success
+        }
+    }
+
+    enum MarkReadError: LocalizedError {
+        case unknown
+    }
+
+    private func notifications(from result: NotificationsResult?) -> Set<NotificationsResult.Notification>? {
+        guard let result = result else {
+            return nil
+        }
+        guard let list = result.query?.notifications?.list else {
+            return nil
+        }
+        return Set(list)
+    }
+
+    public func getAllUnreadNotifications(from subdomains: [String], completion: @escaping (Set<NotificationsResult.Notification>?, Error?) -> Void) {
+        let completion: (NotificationsResult?, URLResponse?, Bool?, Error?) -> Void = { result, _, _, error in
+            guard error == nil else {
+                completion([], error)
+                return
+            }
+            let notifications = self.notifications(from: result)
+            completion(notifications, result?.error)
+        }
+        request(Query.notifications(from: subdomains, limit: .max, filter: .unread), completion: completion)
+    }
+
+    public func markAsRead(_ notifications: Set<RemoteNotification>, completion: @escaping (Error?) -> Void) {
+        let maxNumberOfNotificationsPerRequest = 50
+
+        guard notifications.count <= maxNumberOfNotificationsPerRequest else {
+            // TODO: Split requests? 50 is the limit.
+            assertionFailure()
+            return
+        }
+
+        request(Query.markAsRead(notifications: notifications), method: .post) { (result: MarkReadResult?, _, _, error) in
+            if let error = error {
+                completion(error)
+            }
+            guard let result = result, result.succeeded else {
+                assertionFailure()
+                completion(MarkReadError.unknown)
+                return
+            }
+            completion(result.error)
+        }
+    }
+
+    private func request<T: Decodable>(_ queryParameters: Query.Parameters?, method: Session.Request.Method = .get, completion: @escaping (T?, URLResponse?, Bool?, Error?) -> Void) {
+        let _ = Session.shared.requestWithCSRF(type: CSRFTokenJSONDecodableOperation.self, scheme: NotificationsAPI.scheme, host: NotificationsAPI.host, path: NotificationsAPI.path, method: method, queryParameters: queryParameters, bodyEncoding: .form, tokenContext: CSRFTokenOperation.TokenContext(tokenName: "token", tokenPlacement: .body, shouldPercentEncodeToken: true), completion: completion)
```

Comment:
shouldn't the CSRF token only be required for mutating requests (POST, PUT, PATCH, etc)?

Reply:
Getting all notifications requires a token

Reply:
https://www.mediawiki.org/w/api.php?notwikis=wikidatawiki%7Cenwiki&notlimit=max&action=query&notformat=model&meta=notifications&format=json&formatversion=2&notfilter=!read seems to work for me without a CSRF token

Reply:
maybe it's the cookies, will check 🧐